src/game/mod.rs

```diff
@@ -258,6 +260,79 @@ impl Game {
         Ok((num_cards, drew_monad))
     }
 
+    pub fn check_player_end(&self, player: usize) -> bool {
+        use self::card::Value::*;
+        let player = &self.players[player];
+        if ! (self.table.common.is_empty() || self.table.discard.is_empty() )  {
+            return false;
+        }
+
+        //-------------------------------------------------------------------
+
+        let trade_values: Vec<Value> =
+            (0..player.hand.len())
+                .combinations(2)
+                .filter_map(|pair| player.trade_value(pair[0], pair[1]).ok())
+                .collect();
+
+        if let Some(max_value) = trade_values.iter().max() {
+            if let Some(value) = max_value.succ() {
+                if ! self.table.deck(value).is_empty() {
+                    return false;
+                }
+            }
+            else {
+                return false;
+            }
+        }
+        drop(trade_values);
+
+        //-------------------------------------------------------------------
+
+        let mut sorted_hand = player.hand.clone();
+        let mut sum = 0;
+        let mut highest_value = Common;
+        let mut num_commons = 0;
+
+        sorted_hand.sort_by(|a, b| a.value.cmp(&b.value));
```

Comment:
I would write this as
```rust
sorted_hand.sort_by_key(|card| card.value);
```

Reply:
It was rather late when I wrote this. That is much cleaner.